### PR TITLE
feat: add ability to fail on ERR and other fail-on statuses for breaking changes

### PR DESCRIPTION
**Pull request by jgill-shipwell**
Adds the ability to fail the action/build on other statuses such as `ERR` instead of just `WARN`.

### CODE REVIEW

File: breaking/entrypoint.sh [hunk 1]
```diff
@@ -16,6 +16,8 @@ echo "running oasdiff breaking... base: $base, revision: $revision, fail_on_diff
 flags=""
 if [ "$fail_on_diff" = "true" ]; then
     flags="${flags} --fail-on WARN"
+elif [ "$fail_on_diff" != "false" ]
+    flags="${flags} --fail-on $fail_on_diff"
```
Comment (effoeffi) [inline breaking/entrypoint.sh:20]: I don't think this is a best practice.

Comment (jgill-shipwell) [reply breaking/entrypoint.sh:20]: @effoeffi Is there a way to indicate in the `oasdiff breaking change action` to only `fail` on `ERR` instead of `WARN`?

Comment (effoeffi) [reply breaking/entrypoint.sh:20]: I understand, but it is not best practice to use 'var' as both a boolean and a string. I suggest converting it to a string similar to `oasdiff`. Additionally, update the documentation, defaults, and create tests accordingly. Please also review other parameters for similar improvements.

Comment (jgill-shipwell) [reply breaking/entrypoint.sh]: @effoeffi Updated the `action`, `README`, etc. in  `oasdiff-action` in d96dd2c64ce60803260888512b31a40363c37c68. Which style tests are in the repo or would this be the first tests in the repo if added?

Comment (jgill-shipwell) [reply breaking/entrypoint.sh]: @effoeffi Added a test in https://github.com/oasdiff/oasdiff-action/pull/43/commits/468ff610dbcaff306009a416dc7bf5a849ad40e1.

**Comment by tcooper-uk** [comment]
Taken from the oasdiff help:

```
-o, --fail-on string                 exit with return code 1 when output includes errors with this level or higher: ERR or WARN
```

This means that the behaviour you have described already exists, failing on `WARN` will also fail on `ERR`.

Validated this is true with:
```
oasdiff breaking https://raw.githubusercontent.com/Tufin/oasdiff/main/data/openapi-test1.yaml https://raw.githubusercontent.com/Tufin/oasdiff/main/data/openapi-test4.yaml --fail-on WARN || echo $?
```

**Comment by jgill-shipwell** [comment]
@tcooper-uk Re: https://github.com/oasdiff/oasdiff-action/pull/43#issuecomment-1981695311 We want to fail on `ERR`, but not fail on `WARN` as the warnings are fine in our use case so the `fail-on-diff` that sets the command as `fail-on WARN` did not work for this use case.

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -86,6 +86,31 @@ jobs:
...
+        run: |
+          delimiter=$(cat /proc/sys/kernel/random/uuid | tr -d '-')
+          output=$(cat <<-$delimiter
+          ${{ steps.test_breaking_changes.outputs.breaking }}
+          $delimiter
+          )
+          if [ "$output" != "1 breaking changes: 1 error, 0 warning" ]; then
+            echo "Expected output '6 breaking changes: 2 error, 4 warning' but got '$output'" >&2
+            exit 1
+          fi
```
Comment (effoeffi) [inline .github/workflows/test.yaml:113]: What you are testing?
I suggest taking an example that there is warnings but not errors and set fail-on: ERR and it should not fail

File: README.md [hunk 3]
```diff
@@ -40,13 +40,14 @@ Additional arguments:
...
 | --fail-on WARN            | fail-on-diff            | true    |
+| --fail-on                 | fail-on                 | ''      |
 | --include-checks          | include-checks          | csv     |
 | --include-path-params     | include-path-params     | false   |
 | --deprecation-days-beta   | deprecation-days-beta   | 31      |
 | --deprecation-days-stable | deprecation-days-stable | 180     |
 | --exclude-elements        | exclude-elements        | ''      |
 
-This action delivers a summary of breaking changes, accessible as a GitHub step output named `breaking`.
+This action delivers a summary of breaking changes, accessible as a GitHub step output named `breaking`. Note: `fail-on-diff` takes precedence over `fail-on` when `fail-on-diff` is `true`.
```
Comment (effoeffi) [inline README.md:50]: Lets align with oasdiff by having only `fail-on`. please remove `fail-on-diff`, we can achieve this by using `fail-on: WARN`

File: breaking/entrypoint.sh [hunk 4]
```diff
@@ -4,18 +4,21 @@ set -e
...
+readonly deprecation_days_stable="$8"
+readonly exclude_elements="$9"
 
-echo "running oasdiff breaking... base: $base, revision: $revision, fail_on_diff: $fail_on_diff, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements"
+echo "running oasdiff breaking... base: $base, revision: $revision, fail_on_diff: $fail_on_diff, fail_on: $fail_on, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements"
 
 # Build flags to pass in command
 flags=""
 if [ "$fail_on_diff" = "true" ]; then
     flags="${flags} --fail-on WARN"
```
Comment (effoeffi) [inline breaking/entrypoint.sh:19]: should be removed

Comment (jgill-shipwell) [reply breaking/entrypoint.sh:19]: @effoeffi For https://github.com/oasdiff/oasdiff-action/pull/43#discussion_r1516173477, the reason it was not removed was that it would be a breaking change itself if it was removed from the action. If someone relies on the `fail_on_diff` and it is removed, their workflows will break... of course `fail-on` is more flexible for any status. Will update the `pr` to remove `fail_on_diff` though.

Comment (effoeffi) [reply breaking/entrypoint.sh:19]: > @effoeffi For [#43 (comment)](https://github.com/oasdiff/oasdiff-action/pull/43#discussion_r1516173477), the reason it was not removed was that it would be a breaking change itself if it was removed from the action. If someone relies on the `fail_on_diff` and it is removed, their workflows will break... of course `fail-on` is more flexible for any status. Will update the `pr` to remove `fail_on_diff` though.

@jgill-shipwell that is why we are creating a release, and you should rely on it to avoid any breaking changes.

Comment (jgill-shipwell) [reply breaking/entrypoint.sh]: @effoeffi Updated this `pr` to remove `fail_on_diff` and match the `fail-on` param of `oasdiff`. Updated the tests, etc.

File: breaking/entrypoint.sh [hunk 5]
```diff
@@ -31,12 +31,12 @@ readonly exclude_elements="$8"
...
-echo "running oasdiff breaking... base: $base, revision: $revision, fail_on_diff: $fail_on_diff, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements, composed: $composed, output_to_file: $output_to_file"
+echo "running oasdiff breaking... base: $base, revision: $revision, fail_on: $fail_on, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements"
 
 # Build flags to pass in command
 flags=""
-if [ "$fail_on_diff" = "true" ]; then
-    flags="$flags --fail-on WARN"
+if [ -z "$fail_on" ]; then
+    flags="${flags} --fail-on $fail_on"
 fi
```
Comment (effoeffi) [inline breaking/entrypoint.sh:40]: Do you want to check if it is not empty, right? `-n`?

File: .github/workflows/test.yaml [hunk 6]
```diff
@@ -85,7 +85,7 @@ jobs:
         with:
           base: 'specs/base.yaml'
           revision: 'specs/revision-breaking.yaml'
-          fail-on-diff: false
+          fail-on: 'WARN'
           output-to-file: 'breaking.txt'
```
Comment (effoeffi) [inline .github/workflows/test.yaml:89]: Why this test do not fail?
[oasdiff docs](https://github.com/Tufin/oasdiff/blob/main/BREAKING-CHANGES.md):
> To exit with return code 1 when any breaking changes are found, WARN-level or ERR-level, add the --fail-on WARN flag.

Can you please explain the why you decided to change `fail-on-diff: false` to `fail-on: 'WARN'`?

File: README.md [hunk 7]
```diff
@@ -40,8 +40,8 @@ Copy and paste the following snippet into your build .yml file:
 Additional arguments:
 
 | CLI                       | Action input            | Default |
-| ------------------------- | ----------------------- | ------- |
-| --fail-on WARN            | fail-on-diff            | true    |
+|---------------------------|-------------------------|---------|
+| --fail-on                 | fail-on                 | 'ERR'   |
```
Comment (effoeffi) [inline README.md:44]: The default oasdiff exit code is 0 if there is a breaking changes, why do you want to change that?

File: breaking/entrypoint.sh [hunk 8]
```diff
@@ -31,12 +31,12 @@ readonly exclude_elements="$8"
 readonly composed="$9"
 readonly output_to_file="${10}"
 
-echo "running oasdiff breaking... base: $base, revision: $revision, fail_on_diff: $fail_on_diff, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements, composed: $composed, output_to_file: $output_to_file"
+echo "running oasdiff breaking... base: $base, revision: $revision, fail_on: $fail_on, include_checks: $include_checks, include_path_params: $include_path_params, deprecation_days_beta: $deprecation_days_beta, deprecation_days_stable: $deprecation_days_stable, exclude_elements: $exclude_elements"
```
Comment (effoeffi) [inline breaking/entrypoint.sh:34]: why did you delete `composed: $composed, output_to_file: $output_to_file"`?

Comment (effoeffi) [reply breaking/entrypoint.sh:34]: @jgill-shipwell could you please check (this PR)[https://github.com/oasdiff/oasdiff-action/pull/52] and let me know if it resolves your issue?
